dbms/src/AggregateFunctions/AggregateFunctionSumMap.h

```diff
@@ -0,0 +1,189 @@
+
+        const auto * array_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
+        if (!array_type)
+            throw Exception("First argument for function " + getName() + " must be an array.",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        array_type = checkAndGetDataType<DataTypeArray>(arguments[1].get());
+        if (!array_type)
+            throw Exception("Second argument for function " + getName() + " must be an array.",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        type = arguments.front();
+    }
+
+    void setParameters(const Array & params) override
+    {
+        if (!params.empty())
+            throw Exception("This instantiation of " + getName() + "aggregate function doesn't accept any parameters.",
+                            ErrorCodes::LOGICAL_ERROR);
+    }
+
+    void addImpl(AggregateDataPtr place, const IColumn & column_keys, const IColumn & column_values, size_t row_num, Arena *) const
+    {
+        Field field_keys;
+        column_keys.get(row_num, field_keys);
+        const auto & keys = field_keys.get<Array &>();
+
+        Field field_values;
+        column_values.get(row_num, field_values);
+        const auto & values = field_values.get<Array &>();
+
+        auto & merged_maps = this->data(place).merged_maps;
+
+        if (keys.size() != values.size())
+            throw Exception("Sizes of keys and values arrays do not match", ErrorCodes::LOGICAL_ERROR);
+
+        size_t size = keys.size();
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (merged_maps.find(keys[i]) != merged_maps.end())
+                merged_maps[keys[i]] += values[i];
+            else
+                merged_maps[keys[i]] = values[i];
+        }
+    }
+
+    void merge(AggregateDataPtr place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+        const auto & rhs_maps = this->data(rhs).merged_maps;
+
+        for (const auto &rhs_map : rhs_maps)
+        {
+            if (merged_maps.find(rhs_map.first) != merged_maps.end())
+                merged_maps[rhs_map.first] += rhs_map.second;
+            else
+                merged_maps[rhs_map.first] = rhs_map.second;
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+    {
+        /// Serialize merged_maps as two vectors. Using boost::archive could be better but it's unavailable.
+        const auto & merged_maps = this->data(place).merged_maps;
+        size_t size = merged_maps.size();
+
+        Array keys, values;
+        keys.reserve(size);
+        values.reserve(size);
+        for (const auto &v : merged_maps)
+        {
+            keys.push_back(v.first);
+            values.push_back(v.second);
+        }
+
+        writeVarUInt(size, buf);
+        buf.write(reinterpret_cast<const char *>(&keys[0]), size * sizeof(keys[0]));
```

Comment:
This is not suitable, because:
- arbitary key types are not supported;
- keys will be serialized with garbage.

Better to just write a loop and to use `IDataType::serializeBinary` method.

Reply:
It makes sense to serialize as little as possible. The only question I have - how to get access to data type of keys and values? In method `setArgumentsImpl` I can get access only to data type of arguments itself and that is `DataTypeArray`, but I need to know exact data type of the array elements. Is there a way to do this?

Reply:
You can obtain nested type from DataTypeArray: `DataTypeArray::getNestedType`.

Reply:
Thanks! It worked, new serialization method consumes much less memory now.